go/oasis-node/cmd/debug/txsource/workload/queries.go

```diff
@@ -0,0 +1,455 @@
+	staking "github.com/oasislabs/oasis-core/go/staking/api"
+)
+
+const (
+	// NameQueries is the name of the queries workload.
+	NameQueries = "queries"
+
+	// Ratio of queries that should query height 1.
+	queriesEarliestHeightRatio = 0.1
+)
+
+type queries struct {
+	logger *logging.Logger
+
+	runtimeID       common.Namespace
+	epochtimeParams epochtime.ConsensusParameters
+	stakingParams   staking.ConsensusParameters
+	schedulerParams scheduler.ConsensusParameters
+
+	staking   staking.Backend
+	consensus consensus.ClientBackend
+	registry  registry.Backend
+	scheduler scheduler.Backend
+	runtime   runtimeClient.RuntimeClient
+}
+
+// doConsensusQueries does GetEpoch, GetBlock, GetTransaction queries for the
+// provided height.
+func (q *queries) doConsensusQueries(ctx context.Context, rng *rand.Rand, height int64) error {
+	q.logger.Debug("Doing consensus queries",
+		"height", height,
+	)
+
+	epoch, err := q.consensus.GetEpoch(ctx, height)
+	if err != nil {
+		return fmt.Errorf("GetEpoch at height %d: %w", height, err)
+	}
+	block, err := q.consensus.GetBlock(ctx, height)
+	if err != nil {
+		return fmt.Errorf("GetBlock at height %d: %w", height, err)
+	}
+	if block.Height != height {
+		return fmt.Errorf("block.Height: %d == %d violated", block.Height, height)
+	}
+	if !q.epochtimeParams.DebugMockBackend {
+		expectedEpoch := epochtime.EpochTime(block.Height / q.epochtimeParams.Interval)
+		if expectedEpoch != epoch {
+			q.logger.Error("Invalid epoch",
+				"expected", expectedEpoch,
+				"epoch", epoch,
+				"height", block.Height,
+				"epoch_interval", q.epochtimeParams.Interval,
+			)
+			return fmt.Errorf("Invalid epoch: %d", epoch)
+		}
+	}
+
+	_, err = q.consensus.GetTransactions(ctx, height)
+	if err != nil {
+		return fmt.Errorf("GetTransactions at height %d: %w", height, err)
+	}
+
+	q.logger.Debug("Consensus queries done",
+		"height", height,
+		"epoch", epoch,
+		"block", block,
+	)
+
+	return nil
+}
+
+// doSchedulerQueries does GetCommittees and GetValidator queries for the
+// provided height.
+func (q *queries) doSchedulerQueries(ctx context.Context, rng *rand.Rand, height int64) error {
+	q.logger.Debug("Doing scheduler queries",
+		"height", height,
+	)
+
+	validators, err := q.scheduler.GetValidators(ctx, height)
+	if err != nil {
+		return fmt.Errorf("GetValidators at height %d: %w", height, err)
+	}
+	if len(validators) < q.schedulerParams.MinValidators {
+		return fmt.Errorf("Not enough validators at height %d, expected at least: %d, got: %d", height, q.schedulerParams.MinValidators, len(validators))
+	}
+
+	// Query commiteess for the runtime.
+	committees, err := q.scheduler.GetCommittees(ctx, &scheduler.GetCommitteesRequest{
+		Height:    height,
+		RuntimeID: q.runtimeID,
+	})
+	if err != nil {
+		return fmt.Errorf("GetCommittees at height %d: %w", height, err)
+	}
+	// First two epochs shouldn't have runtime committees, following epochs
+	// should all have existing committess for the simple-keyvalue runtime.
```

Comment:
In the E2E-tests we currently have only validators in the genesis documents, meaning there will be no other committees in the first epoch (epoch=0). In epoch 0 the key-manager will register, but not yet compute/storage workers, since those are waiting for key-manager committee to be available, see: https://github.com/oasislabs/oasis-core/blob/7c800a8cf4c0350de769b59e45b0547499a8dce1/go/worker/common/committee/node.go#L291 (which will happen at start of epoch 1). 
This means that compute/storage nodes will get registered during epoch 1 and the committee should be elected from epoch 2 onward.

Maybe i should explain this in the comment, and move the threshold epoch at which committee is expected into a constant.

---

However indeed below should `if` should be changed to `if epoch > 1`.

Reply:
thanks for the elaboration